### ISSUE TITLE
invalid source map if css:false

### ISSUE THREAD

**Issue by apairon** [EMH333/esbuild-svelte#13]
Hello,

I use esbuild with this plugin and I have found the following problem if I enable source map generation with `css` set to `false` in `compileOptions`

```javascript
const esbuild_svelte = require("esbuild-svelte/configurable")({
    compileOptions: { css: false, hydratable: true },
    preprocessor: svelteConfig.preprocess,
}
```
I think the source of the problem is in your code at `configurable/index.js` line 44:

```javascript
                    //if svelte emits css seperately, then store it in a map and import it from the js
                    if (!compileOptions.css && css.code) {
                        let cssPath = args.path.replace(".svelte", ".esbuild-svelte-fake-css").replace(/\\/g, "/");
                        cssCode.set(cssPath, css.code + `/*# sourceMappingURL=${css.map.toUrl()}*/`);
                        contents = `import "${cssPath}";\n` + contents;
                    }
```

If I remove the `\n` the source map line is ok:

```javascript
...
                        contents = `import "${cssPath}";` + contents;
...
```

I don't know if this will fix the problem completely. Possibly there is another way in esbuild to inject the css without modifying the contents.

**Comment by EMH333** [comment]
Ah, I'm guessing you are seeing the source map be off by one line?

I will switch it so the import is at the bottom of the file instead. Thanks for the report!

**Comment by EMH333** [comment]
@apairon Just released `v0.3.2` which should fix this issue. Can you please confirm?

**Comment by apairon** [comment]
`v0.3.2` fixes the problem with the wrong ling, but the source is the already preprocessed source from `svelte-preprocess`. So no typescript is shown in my case. I don't know if it is a missing feature in svelte/compiler, svelte-preprocess or a setting in your plugin. But for now it is ok for me and better than having no source map. Thank you

**Comment by EMH333** [comment]
Looks like the Svelte project is in the process of figuring out how to deal with preprocessor sourcemaps. I'll keep this issue open (and keep an eye on their progress) till everything settles down a bit

**Comment by EMH333** [comment]
I have figured out how to support preprocessor sourcemaps, however I had to kluge my way around the way Svelte has them implemented so it isn't quite ready for release yet. I'm hoping to have a fully tested implementation in place and released by the end of the week

**Comment by EMH333** [comment]
@apairon Is it possible for you to test the changes introduced in 103ef96ca9b3d4357ff6cb15cea8cdb921cf7d02? If not, no problem just thought I would check before releasing to the world.

It should be in a pretty good place, but there is one path that I couldn't find a way to test and I want to double check that it won't cause problems when I release it.

**Comment by apairon** [comment]
@EMH333 I tested it with my project, but it failed with an error for every svelte import, pe.:

```
 > src/components/routes/checkout/Overview.svelte: warning: [esbuild-svelte] There was an error while dealing with the preprocessor sourcemap in the esbuild-svelte plugin. Please file a bug at https://github.com/EMH333/esbuild-svelte/issues detailing your use case (types of content being preprocessed and preprocessors used) so the issue can be fixed promptly
    37 │ import Cart from "../../layout/Cart.svelte";
```

Output of `console.log(js.map.sources)` in front of the error printing in your code:

```js
[ 'src/components/layout/Cart.svelte', 'Cart.svelte' ]
```

For every import statement of svelte files there are 2 items in the `js.map.sources` array so that your warning message is triggered.

Source mapping for svelte files is now completly wrong. It always maps to my index.html file. `*.ts` files mapping is correct (I know it is not touched by your plugin, but for checking esbuild correctness).

**Comment by EMH333** [comment]
I obviously have more work to do, but I think I have a better idea of what is going wrong. Thank you for testing!

**Comment by EMH333** [comment]
Can you please try again (this time with the `preprocessor-sourcemaps` branch)? Also, please make sure you are using the latest version of `esbuild` and `svelte-preprocess`. In the last few releases there have been significant improvements that should make an impact. Another thing to try is to install `svelte` as a dev dependency as well, to make sure the latest version is being used.

I suspect your duplication of files in the sourcemap is actually an `esbuild` related thing and I haven't been able to replicate it when using newer versions.

Thank you for your help!

**Comment by apairon** [comment]
I tested with the branch `preprocessor-sourcemaps` and there are the same errors. Not every `js.map.sources` has 2 items, but most of them.

Here ist my `yarn list | egrep  "svelte|esbuild"` after `yarn upgrade`:

```
├─ @tsconfig/svelte@1.0.10
├─ esbuild-svelte@0.4.0
│  └─ svelte@^3.31.0
├─ esbuild@0.8.36
├─ prettier-plugin-svelte@1.4.2
├─ svelte-check@1.1.31
│  ├─ svelte-preprocess@^4.0.0
├─ svelte-hmr@0.11.6
├─ svelte-lazy@1.0.8
├─ svelte-loading-spinners@0.1.4
├─ svelte-material-icons@1.0.3
├─ svelte-materialify@0.3.4
├─ svelte-preprocess-esbuild@1.0.4
├─ svelte-preprocess@4.6.5
├─ svelte-routing@1.5.0
├─ svelte-scrollto@0.2.0
│  └─ svelte@3.x
├─ svelte@3.32.0
```

Later I will try it with a new project. Now I see no relation between the import statement and the output of one or two items in the `js.map.sources` array.

I think my current project is too complex for tests. Starting with a new small project hopefully I can discover when the duplication of files happen.

**Comment by EMH333** [comment]
**I think I have figured out why this is happening. Can you please list the preprocessors that you are using to confirm?** 

(for my use):
If the `sources` entry is different for the same file, then it's treated as a totally different file. As a result, it breaks my code because it is only expecting one file output. 
```js
SourceMap {
  version: 3,
  mappings: 'AAAA,CAAC,MAAM,CAAC,IAAI,CAAC,CAAC,UAAU,CAAC,CCKrB,QAAA,CAAA,KAAA,CAAA,CAAA,CAAiB;AACb,EAAA,YAAA,CAAa,CAAA,CAAE,aAAA,CAAA,CAAe,CAAA,SAAA,CAAA,CAAA,CAAA,CAAA,CAAA;AAC9B,EAAA,MAAA,CAAA,CAAA;AAAA;AAGJ,QAAA,CAAA,YAAA,CAAsB,KAAA,CAAA,CAAa;AAC/B,EAAA,OAAA,CAAQ,GAAA,CAAI,KAAA,CAAM,aAAA,CAAc,WAAA,CAAA,CAAA,CAAA;AAAA;AAGpC,GAAA,CAAI,YAAA;AAEJ,QAAA,CAAA,MAAA,CAAA,CAAA,CAAkB;AACd,EAAA,YAAA,CAAa,KAAA,CAAA,CAAA,CAAQ,CAAA,OAAA,CAAA,CAAA,CAAA,CAAY,IAAA,CAAK,KAAA,CAAM,IAAA,CAAK,MAAA,CAAA,CAAA,CAAA,CAAA,CAAW,GAAA,CAAA;AAAA;ADEpE,CAAC,CAAC,MAAM;;AAER,CAAC,KAAK,CAAC,IAAI,CAAC,CAAC,IAAI,CAAC;;;;;;;CAQlB,CAAC,CAAC,KAAK;;AAEP,CAAC,MAAM,CAAC,EAAE,CAAC,KAAK,CAAC,CAAC,KAAK,CAAC,CAAC,KAAK,CAAC,CAAC,SAAS,CAAC,CAAC,IAAI,CAAC,CAAC,SAAS,CAAC,CAAC,CAAC,MAAM;AACnE,CAAC,KAAK,CAAC,IAAI,CAAC,CAAC,IAAI,CAAC,CAAC,IAAI,CAAC,CAAC,YAAY,CAAC,CAAC,IAAI,CAAC,IAAI,CAAC,CAAC,YAAY,CAAC,CAAC,CAAC;',
  names: [],
  sources: [
    'pp-sourcemaps.svelte',
    'test/fixtures/preprocessing-sourcemaps/pp-sourcemaps.svelte'
  ]
}
```
In this case the two plugins were `svelte-preprocess-sass` and `svelte-preprocess-esbuild`.
Need to find some way of combining those sourcemaps if same file (and figuring out if they are the same file reliably). It's possible that preprocessors will return multiple files so will need to deal with that in a graceful way. Should have test for this case, one source and multiple sources of different files.

TL;DR: Some preprocessors aren't using the same source name for the input file. This causes problems on my end

**Comment by apairon** [comment]
Currently I use:
```javascript
const { typescript } = require("svelte-preprocess-esbuild")
const sveltePreprocess = require("svelte-preprocess")
module.exports = { 
    preprocess: [ // importing this to: preprocess attribute (esbuild-svelte) in other file
        typescript({
            sourcemap: true,
        }),
        sveltePreprocess({
            sourceMap: true,
            typescript: false,
            /* scss: {
                includePaths: ["src/theme"],
            }, */
        }),
    ],
}
```

I can use the current version of esbuild-svelte 0.5.4 without warnings or errors in my test project. I don't know in which version it fixed my problem because I used an old version of esbuild-svelte in my other project. Now the only problem is still the referenced source in browser is after preprocessing.

See following test project:

Original:
```svelte
<h1>{title}</h1>

<div use:links>
    <a href="/test1">1</a>
    <a href="/test2">2</a>
    <a href="/test3">3</a>
    <a href="/test4">4</a>
    <Router url="{url}">
        <Route path="/*path" let:params>
            <Content path="/{params.path}" />
        </Route>
    </Router>
</div>

<style lang="less">
    @import "./../css/main.less";
    h1 {
        color: red;
    }

</style>

<script lang="typescript">
    import { Router, Route, links } from "svelte-routing"
    import { scrollToTop } from "svelte-scrollto"
    import { title } from "../config"
    import { location } from "../store"
    import Content from "./Content.svelte"

    export let url = ""
    if (url) {
        // ssr
        let l = url.split("?")
        $location = {
            path: l[0],
            search: l.length > 1 ? l[1] : "",
            categoryPath: l[0].replace(/\/\d{4,99}[^\/]+$/, ""),
            push: false,
            pop: false,
        }
    }

    // scroll to top on new site
    location.subscribe((l) => {
        if (l.push) scrollToTop()
    })

    if (typeof window !== "undefined") console.log("App initialized")

</script>
```

Console.log says:
```
App initialized        App.svelte:41
```
and is referencing in browser to:

```svelte
<h1>{title}</h1>

<div use:links>
    <a href="/test1">1</a>
    <a href="/test2">2</a>
    <a href="/test3">3</a>
    <a href="/test4">4</a>
    <Router url="{url}">
        <Route path="/*path" let:params>
            <Content path="/{params.path}" />
        </Route>
    </Router>
</div>

<style lang="less">h1 {
  color: red;
}
</style>

<script lang="typescript">import { Router, Route, links } from "svelte-routing";
import { scrollToTop } from "svelte-scrollto";
import { title } from "../config";
import { location } from "../store";
import Content from "./Content.svelte";
let url = "";
if (url) {
  let l = url.split("?");
  $location = {
    path: l[0],
    search: l.length > 1 ? l[1] : "",
    categoryPath: l[0].replace(/\/\d{4,99}[^\/]+$/, ""),
    push: false,
    pop: false
  };
}
location.subscribe((l) => {
  if (l.push)
    scrollToTop();
});
if (typeof window !== "undefined")
  console.log("App initialized");
export {
  url
};
</script>
```

**Comment by apairon** [comment]
I did some experiments and found a way to get correct sourcemaps with preprocessor.

If you capture the `preprocessResult.map` and add it to `compileOptions.sourcemap` you get a correct sourcemap. The problem is now that the `sourcesContent` in the final map-object is not set and `sources`-paths after esbuild pe. instead of `src/components...`  they are `src/components/src/components/...`. I fixed this via removing the path and only using filenames in `sources` and filling the content of the files into `sourcesContent`.

I don't know if it is the correct way to push sourcemaps to the svelte compiler, but for my project it works. Sorry for not forking and doing a pull request. But these are only experiments without understanding the the sourcemap handling of svelte compiler.

Maybe it is a starting point for you to implement preprocessor sourcemap handling.

Here is your modified code (ts->js and some mods):

```javascript
const { preprocess, compile } = require("svelte/compiler")
const { dirname, relative } = require("path")
const { promisify } = require("util")
const { readFile, statSync, readFileSync } = require("fs")

const convertMessage = ({ message, start, end, filename, frame }) => ({
    text: message,
    location: start &&
        end && {
            file: filename,
            line: start.line,
            column: start.column,
            length: start.line === end.line ? end.column - start.column : 0,
            lineText: frame,
        },
})

const SVELTE_FILTER = /\.svelte$/
const FAKE_CSS_FILTER = /\.esbuild-svelte-fake-css$/

module.exports = (options) => {
    return {
        name: "esbuild-svelte",
        setup(build) {
            if (!options) {
                options = {}
            }
            // see if we are incrementally building or watching for changes and enable the cache
            // also checks if it has already been defined and ignores this if it has
            if (
                options.cache == undefined &&
                (build.initialOptions.incremental || build.initialOptions.watch)
            ) {
                options.cache = true
            }

            // disable entry file generation by default
            if (options.fromEntryFile == undefined) {
                options.fromEntryFile = false
            }

            //Store generated css code for use in fake import
            const cssCode = new Map()
            const fileCache = new Map()

            //check and see if trying to load svelte files directly
            build.onResolve({ filter: SVELTE_FILTER }, ({ path, kind }) => {
                if (kind === "entry-point" && options?.fromEntryFile) {
                    return { path, namespace: "esbuild-svelte-direct-import" }
                }
            })

            //main loader
            build.onLoad(
                {
                    filter: SVELTE_FILTER,
                    namespace: "esbuild-svelte-direct-import",
                },
                async (args) => {
                    return {
                        errors: [
                            {
                                text:
                                    "esbuild-svelte doesn't support creating entry files yet",
                            },
                        ],
                    }
                }
            )

            //main loader
            build.onLoad({ filter: SVELTE_FILTER }, async (args) => {
                // if told to use the cache, check if it contains the file,
                // and if the modified time is not greater than the time when it was cached
                // if so, return the cached data
                if (options?.cache === true && fileCache.has(args.path)) {
                    const cachedFile = fileCache.get(args.path) || {
                        dependencies: new Map(),
                        data: null,
                    } // should never hit the null b/c of has check
                    let cacheValid = true

                    //for each dependency check if the mtime is still valid
                    //if an exception is generated (file was deleted or something) then cache isn't valid
                    try {
                        cachedFile.dependencies.forEach((time, path) => {
                            if (statSync(path).mtime > time) {
                                cacheValid = false
                            }
                        })
                    } catch {
                        cacheValid = false
                    }

                    if (cacheValid) {
                        return cachedFile.data
                    } else {
                        fileCache.delete(args.path) //can remove from cache if no longer valid
                    }
                }

                //reading files
                let source = await promisify(readFile)(args.path, "utf8")
                let filename = relative(process.cwd(), args.path)

                //file modification time storage
                const dependencyModifcationTimes = new Map()
                dependencyModifcationTimes.set(
                    args.path,
                    statSync(args.path).mtime
                ) // add the target file

                //actually compile file



// MOD START
                let preprocessMap
// MOD END




                try {
                    //do preprocessor stuff if it exists
                    if (options?.preprocess) {
                        let preprocessResult = await preprocess(
                            source,
                            options.preprocess,
                            {
                                filename,
                            }
                        )
                        source = preprocessResult.code


// MOD START
                        preprocessMap = preprocessResult.map
// MOD END




                        // if caching then we need to store the modifcation times for all dependencies
                        if (options?.cache === true) {
                            preprocessResult.dependencies?.forEach((entry) => {
                                dependencyModifcationTimes.set(
                                    entry,
                                    statSync(entry).mtime
                                )
                            })
                        }
                    }

                    let compileOptions = {
                        css: false,
                        ...options?.compileOptions,
                    }




// MOD START
// - remove leading paths, so that final paths after esbuild in sourcemap are correct in browser log
// - add map to compiler options
                    if (preprocessMap) {
                        for (let i = 0; i < preprocessMap.sources.length; i++) {
                            preprocessMap.sources[i] = preprocessMap.sources[
                                i
                            ]?.replace(/(.+\/)+/, "")
                        }
                        compileOptions.sourcemap = preprocessMap
                    }
// MOD END





                    let { js, css, warnings } = compile(source, {
                        ...compileOptions,
                        filename,
                    })




// MOD START
// - add original file content to source map, other way is to provide source files via web server
                    if (!js.map.sourcesContent) {
                        try {
                            js.map.sourcesContent = [
                                readFileSync(filename, "utf8"),
                            ]
                        } catch (e) {}
                    }
// MOD END





                    let contents =
                        js.code + `\n//# sourceMappingURL=` + js.map.toUrl()

                    //if svelte emits css seperately, then store it in a map and import it from the js
                    if (!compileOptions.css && css.code) {
                        let cssPath = args.path
                            .replace(".svelte", ".esbuild-svelte-fake-css")
                            .replace(/\\/g, "/")
                        cssCode.set(
                            cssPath,
                            css.code +
                                `/*# sourceMappingURL=${css.map.toUrl()} */`
                        )
                        contents = contents + `\nimport "${cssPath}";`
                    }

                    const result = {
                        contents,
                        warnings: warnings.map(convertMessage),
                    }

                    // if we are told to cache, then cache
                    if (options?.cache === true) {
                        fileCache.set(args.path, {
                            data: result,
                            dependencies: dependencyModifcationTimes,
                        })
                    }

                    // make sure to tell esbuild to watch any additional files used if supported
                    if (build.initialOptions.watch) {
                        // this array does include the orignal file, but esbuild should be smart enough to ignore it
                        result.watchFiles = Array.from(
                            dependencyModifcationTimes.keys()
                        )
                    }

                    return result
                } catch (e) {
                    return { errors: [convertMessage(e)] }
                }
            })

            //if the css exists in our map, then output it with the css loader
            build.onResolve({ filter: FAKE_CSS_FILTER }, ({ path }) => {
                return { path, namespace: "fakecss" }
            })

            build.onLoad(
                { filter: FAKE_CSS_FILTER, namespace: "fakecss" },
                ({ path }) => {
                    const css = cssCode.get(path)
                    return css
                        ? {
                              contents: css,
                              loader: "css",
                              resolveDir: dirname(path),
                          }
                        : null
                }
            )
        },
    }
}

```

By the way: The above code works well if preprocessor is pe.:

```javascript
    preprocess: [
        sveltePreprocess({
            sourceMap: true,
        }),
    ],
```

But if the preprocessor is pe.:

```javascript
    preprocess: [
        typescript({
            sourcemap: true,
        }),

        sveltePreprocess({
            sourceMap: true,
            typescript: false,
        }),
    ],
```

then it only works if svelte files have the layout `scripts - styles - markup` and not `markup - styles - scripts`. With markup in front the offset is incorrect. Line 1 is then also starting in script tag for typescript scripts.

**Comment by EMH333** [comment]
Good to see some additional investigation! The best I can tell, the way Svelte preprocessors handle the path is inconsistent and sometimes results in multiple duplication of the same file in the source map. Your modifications are very similar to those that I had in mind, though I'm not sure they work if a preprocessor returned multiple files (for example importing a stylesheet in scss) which is something that should be covered in the final version. I would need to test to confirm.

You brought up an interesting point: 
>  then it only works if svelte files have the layout `scripts - styles - markup` and not `markup - styles - scripts`. With markup in front the offset is incorrect. Line 1 is then also starting in script tag for typescript scripts.  
 
Which will also need to be checked, so thank you!

I'm slowly closing in on a way to fix this (even if it doesn't appear that way publicly) and plan to get some sort of test patch out soon (within the month). I also have been meaning to check on how other svelte-<bundler> plugins handle this because it seems like something they would run into. Thanks again, and I'll keep this updated with any additional progress
